File: leetcode/cpp/p45-jump-game-ii.cpp
// LeetCode 45. Jump Game II.
#include <iostream>
#include <vector>

#include "leetcode.hpp"

class Solution {
public:
    int jump(const std::vector<int>& nums)
    {
        size_t ans = 0;
        size_t farthest = 0;
        size_t end = 0;
        for (size_t i = 0; i < nums.size() - 1; ++i) {
            farthest = std::max(farthest, i + nums[i]);
            if (farthest >= nums.size() - 1) {
                ++ans;
                break;
            }
            if (i == end) {
                ++ans;
                end = farthest;
            }
        }
        return ans;
    }
};

int main()
{
    struct TestCase {
        std::vector<int> nums;
        int exp;
    };
    Solution solution;
    const TestCase test_cases[] = {
        { { 5 }, 0 },
        { { 1, 1 }, 1 },
        { { 1, 0 }, 1 },
        { { 1, 1, 7 }, 2 },
        { { 2, 1, 7 }, 1 },
        { { 2, 3, 1, 1, 4 }, 2 },
        { { 2, 3, 0, 1, 4 }, 2 },
        { { 5, 0, 0, 0, 0, 0 }, 1 },
    };
    for (const auto& tc : test_cases) {
        const auto act = solution.jump(tc.nums);
        if (tc.exp != act) {
            std::cerr << "FAIL, "
                      << "nums: " << leetcode::to_string(tc.nums)
                      << ", exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
    return 0;
}

File: leetcode/rust/src/p45-jump-game-ii.rs
// LeetCode 45. Jump Game II.

use std::vec::Vec;

struct Solution {}

impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let mut ans: i32 = 0;
        let mut farthest: usize = 0;
        let mut end: usize = 0;
        for (i, n) in nums[0..nums.len() - 1].into_iter().enumerate() {
            farthest = std::cmp::max(farthest, i + *n as usize);
            if farthest >= nums.len() - 1 {
                ans += 1;
                break;
            }
            if i == end {
                ans += 1;
                end = farthest;
            }
        }
        return ans;
    }
}

fn main() {
    struct TestCase {
        nums: Vec<i32>,
        exp: i32,
    }

    let test_cases = vec![
        TestCase {
            nums: vec![5],
            exp: 0,
        },
        TestCase {
            nums: vec![1, 1],
            exp: 1,
        },
        TestCase {
            nums: vec![1, 0],
            exp: 1,
        },
        TestCase {
            nums: vec![1, 1, 7],
            exp: 2,
        },
        TestCase {
            nums: vec![2, 1, 7],
            exp: 1,
        },
        TestCase {
            nums: vec![2, 3, 1],
            exp: 1,
        },
        TestCase {
            nums: vec![2, 3, 1, 1, 4],
            exp: 2,
        },
        TestCase {
            nums: vec![2, 3, 0, 1, 4],
            exp: 2,
        },
        TestCase {
            nums: vec![5, 0, 0, 0, 0],
            exp: 1,
        },
    ];

    for tc in test_cases.iter() {
        let act = Solution::jump(tc.nums.clone());
        if tc.exp != act {
            println!("FAIL. nums: {:?}, exp: {}, act: {}", tc.nums, tc.exp, act);
        }
    }
}
